src/main/java/org/dave/compactmachines3/world/tools/TeleportationTools.java

```diff
@@ -54,29 +80,63 @@ public static boolean tryToEnterMachine(EntityPlayer player, TileEntityMachine m
         if (isNew) // A new grid position was chosen if the id had not yet been set before the structure init, have to update for clients
             MessageMachinePositions.updateClientMachinePositions();
 
-        TeleportationTools.teleportPlayerToMachine((EntityPlayerMP) player, machine);
+        WorldServer machineWorld = DimensionTools.getServerMachineWorld();
+        BlockPos destination = new BlockPos(WorldSavedDataMachines.getInstance().spawnPoints.get(machine.id).getPosition());
+
+        BlockPos safe = null;
+        if (!ConfigurationHandler.MachineSettings.allowUnsafeTeleport && (!machineWorld.isAirBlock(destination) || !machineWorld.isAirBlock(destination.add(0, 1, 0)))) {
+            safe = TeleportationTools.findSafeSpawnpoint(machine);
+            if (safe == null) {
+                ITextComponent message = new TextComponentTranslation("hint.compactmachines3.no_spawnpoint_found")
+                        .setStyle(new Style().setColor(TextFormatting.RED));
+                player.sendStatusMessage(message, true);
+                return false;
+            }
+        }
+
+        machineExitData.remove(player.getUniqueID()); // Remove old exit data
+        TeleportationTools.teleportPlayerToMachine(player, machine.id, false, safe);
+        if (safe != null) {
+            ITextComponent message = new TextComponentTranslation("hint.compactmachines3.used_safe_spawnpoint")
+                    .setStyle(new Style().setColor(TextFormatting.GREEN));
+            player.sendStatusMessage(message, true);
+        }
```

Comment:
Comment this section better. It appears to be teleport handled, but there are a lot of method calls that can be documented.